cc/model/model.h

```diff
@@ -28,33 +28,84 @@ namespace minigo {
 
 class Model {
  public:
-  struct Input {
-    // Whether it's Black or White to play.
-    Color to_play;
+  enum class FeatureType {
+    kAgz,
+    kExtra,
+
+    kNumFeatureTypes,
+  };
+
+  static constexpr int kNumAgzFeaturePlanes = 17;
+  static constexpr int kNumExtraFeaturePlanes = 20;
+
+  // A simple tensor representation that abstracts a real engine-specific
+  // tensor. Tensor does not own the memory pointed to by `data`.
+  // Tensors are assumed to be tightly packed for now.
+  // TODO(tommadams): Make this templated on the data type so we can support
+  // byte input features, and quantized outputs.
+  template <typename T>
+  struct Tensor {
+    Tensor() = default;
+    Tensor(int n, int h, int w, int c, T* data)
+        : n(n), h(h), w(w), c(c), data(data) {}
+
+    int n = 0;
+    int h = 0;
+    int w = 0;
+    int c = 0;
+    T* data = nullptr;
+  };
 
+  template <typename T>
+  class BackedTensor {
+   public:
+    BackedTensor() = default;
+    BackedTensor(int n, int h, int w, int c) {
+      resize(n, h, w, c);
+    }
+
+    void resize(int n, int h, int w, int c) {
+      auto size = n * h * w * c;
+      if (static_cast<size_t>(size) > buffer_.size()) {
+        buffer_.resize(size);
+      }
+      tensor_ = {n, h, w, c, buffer_.data()};
+    }
+
+    operator Tensor<T>&() { return tensor_; }
+    operator const Tensor<T>&() const { return tensor_; }
+
+   private:
+    Tensor<T> tensor_;
+    std::vector<T> buffer_;
+  };
+
+  struct Input {
     // Symmetry to apply to the input features when performing inference.
-    symmetry::Symmetry sym;
+    symmetry::Symmetry sym = symmetry::kNumSymmetries;
```

Comment:
isn't this out of bounds?  should be kIdentity?

Reply:
Yes. That's intentional so that we'll immediately check fail in `ApplySymmetry` if it's not set (rather than silently using `kIdentity`: https://github.com/tensorflow/minigo/blob/9fc09ee0ad5a061eda9c19ace5b163caf8513d5f/cc/symmetries.h#L195